#summary A quick tour using a small sample data set

<wiki:toc max_depth="2"/>

= Introduction =
The following section provides a quick and fast introduction into the analysis of GO enrichment of candidate SNPs using Gowinda. For the sake of this tutorial we provide a small artificial data set for _Drosophila melanogaster_ chromosome 2R. In these sample data we identified 348.701 SNPs and a GWAS-like study showed that 279 of those are associated with the trait of interest, lets say wing length. We furthermore obtained an annotation for _D. melanogaster_ from FlyBase (http://flybase.org/) and the GO associations from the FuncAssociate2 Download section (http://llama.mshri.on.ca/funcassociate/). In the following section we demonstrate how to test, whether the set of candidate SNPs shows a significant enrichment for any GO category. 


= Data =
The following sample data set contains SNPs, annotation and the GO associations for _D.melanogaster_ chromosome arm 2R. 
First obtain the sample data for the Walkthrough: http://gowinda.googlecode.com/files/GowindaWalkthrough.zip

After unzipping the archive you should have the following files:
  * all_snps.txt: the set of all identified SNPs
  * candidate_snps.txt: the candidate SNPs. Some previous GWAS-like analysis revealed that these SNPs are associated with a trait of interest for example wing length. They are a subset of all_snps.txt
  * annotation.gff: Annotation of the genome in the .gff file format (obtained from !FlyBase: http://flybase.org/)
  * funcassociate_go_associations.txt: A GO annotation file obtained from the !FuncAssociate2 web page (http://llama.mshri.on.ca/funcassociate_client/html/) 




= Walkthrough =

== Gene based analysis ==